File: typescript-javascript/algorithms/easy/countingValleys.ts
/**
 * @author Rithie Natan Carvalhaes Prado
 * @date 2026-05-19
 * Last modified: 2026-06-11
 * @description Counting Valleys - HackerRank
 * @see {@link https://www.hackerrank.com/challenges/counting-valleys/problem}
 */
//---------- functions ----------
/**
 * Counts the number of valleys in a hiking trail.
 * @param {number} steps - The number of steps in the trail.
 * @param {string} path - A string representing the trail, where 'U' indicates an uphill step and 'D' indicates a downhill step.
 * @returns {number} - The number of valleys in the trail.
 */
function countingValleys(steps: number, path: string): number 
{
    let starts = path[0], last = "";
    let sealvl = path[0] === "D" ? -1 : 1;
    let valleys = 0;
    
    for(let i = 1; i < steps; i++)
    {
        if(path[i] === "D")
        { sealvl--; }
        else
        { sealvl++; }
        last = path[i];
        
        if(sealvl === 0 && starts === "D" && last === "U")
        { valleys++; starts = path[i+1]; last = ""; i++; sealvl = path[i] === "D" ? -1 : 1; }
        else if(sealvl === 0 && starts === "U" && last === "D")
        { starts = path[i+1]; last = ""; i++; sealvl = path[i] === "D" ? -1 : 1; }
    }//end for

    return(valleys);
}//end countingValleys()

File: python/algorithms/easy/countingValleys.py
"""
Author: Rithie Natan Carvalhaes Prado
Date: 2026-06-13
Last modified: 2026-06-13
Description: Counting Valleys - HackerRank
Test description: https://www.hackerrank.com/challenges/counting-valleys/problem
"""
#---------- functions ----------
def countingValleys(steps: int, path: str) -> int:
    """
    Counts the number of valleys in a given path.

    Parameters:
        steps (int): The number of steps in the path.
        path (str): A string representing the path, where 'U' is an uphill step and 'D' is a downhill step.

    Returns:
        int: The number of valleys in the path.
    """
    
    starts = path[0]
    last = ""
    seaLevel = -1 if path[0] == "D" else 1
    valleys = 0
    i = 1

    while i < steps:
        if path[i] == "D":
            seaLevel -= 1
        else:
            seaLevel += 1

        last = path[i]

        if seaLevel == 0 and starts == "D" and last == "U":
            valleys += 1
            i += 1
            last = ""

            if i <= steps-2:
                starts = path[i]
                seaLevel = -1 if path[i] == "D" else 1
        elif seaLevel == 0 and starts == "U" and last == "D":
            starts = path[i]
            last = ""
            i += 1
            
            if i < steps-2:
                seaLevel = -1 if path[i] == "D" else 1

        i += 1

    return valleys